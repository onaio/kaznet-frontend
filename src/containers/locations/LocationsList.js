// Smart component that renders the Locations List view
import React, { Component } from "react";
import voca from "voca";
import { connect } from "react-redux";
import { Link } from "react-router-dom";
import { bindActionCreators } from "redux";
import qs from "qs";
import * as locationActions from "../../store/locations/actions";
import * as locationSelectors from "../../store/locations/reducer";
import * as globalActions from "../../store/global/actions";
import * as globalSelectors from "../../store/global/reducer";
import * as constants from "../../constants.js";
import "../LoadListAnimation.css";
import ListView from "../../components/ListView";
import NotFound from "../../components/NotFound";
import ElementMap from "../ElementMap";

export class LocationsList extends Component {
  async componentDidMount() {
    this.props.showListTitle();
    this.props.changePageTitle("Locations");
    this.props.changePageTitleButton("+ Create Location");
    this.props.changePageTarget("/locations/new");

    let { search } = qs.parse(this.props.location.search.slice(1));
    const { page } = qs.parse(this.props.location.search.slice(1));

    if (search === undefined) {
      search = "";
    }
    this.props.searchVal(search);
    let pageNumber = Number(page);

    if (isNaN(pageNumber)) {
      pageNumber = 1;
    }

    await this.props.fetchLocations(
      `${constants.API_ENDPOINT}/locations/?search=${search}&page=${pageNumber}`
    );
    this.props.changePageNumber(pageNumber);
  }

  componentDidUpdate(prevProps) {
    let { search } = qs.parse(this.props.location.search.slice(1));
    if (search === undefined) {
      search = "";
    }
    const { page } = qs.parse(this.props.location.search.slice(1));
    if (Number(page) !== this.props.currentPage && !isNaN(page)) {
      const pageNumber = Number(page);
      this.props.fetchLocations(
        `${
          constants.API_ENDPOINT
        }/locations/?search=${search}&page=${pageNumber}`
      );
      this.props.changePageNumber(pageNumber);
    }
  }

  render() {
    if (this.props.searchParam !== "" && this.props.rowsIdArray.length === 0) {
      return <NotFound searchVal={this.props.searchParam} />;
    }
    if (this.props.rowsIdArray.length <= 0) return this.renderLoading();
    return (
      <div className="LocationList">
        <ListView
          renderHeaders={this.renderHeaders}
          rowsIdArray={this.props.rowsIdArray}
          rowsById={this.props.rowsById}
          renderRow={this.renderRow}
          endpoint={"locations"}
          pageLinks={this.props.pageLinks}
          totalPages={this.props.totalPages}
          currentPage={this.props.currentPage}
          firstPage={this.props.firstPage}
          lastPage={this.props.lastPage}
          searchVal={this.props.searchParam}
        />
      </div>
    );
  }

  renderLoading() {
    return (
      <center>
        <div className="lds-ripple">
          <div />
          <div />
        </div>
      </center>
    );
  }

  renderRow(row) {
    const rowItems = [
      <Link to={`/locations/${row.id}`} key="link_to">
        {row.attributes.name}
      </Link>,
      row.attributes.parent_name,
      row.attributes.location_type_name,
      voca.truncate(row.attributes.description, 60)
    ];
    return <ElementMap items={rowItems} HTMLTag="td" />;
  }

  renderHeaders() {
    const headerItems = ["Name", "Parent", "Type", "Description"];
    return <ElementMap items={headerItems} HTMLTag="th" />;
  }
}

function mapStateToProps(state) {
  return {
    rowsById: locationSelectors.getLocationsById(state),
    rowsIdArray: locationSelectors.getLocationsIdArray(state),
    totalPages: locationSelectors.getTotalPages(state),
    currentPage: locationSelectors.getCurrentPage(state),
    pageLinks: locationSelectors.getPageLinks(state),
    firstPage: locationSelectors.getFirstPage(state),
    lastPage: locationSelectors.getLastPage(state),
    searchParam: globalSelectors.getSearchValue(state)
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    {
      fetchLocations: locationActions.fetchLocations,
      changePageNumber: locationActions.changePageNumber,
      changePageTitle: globalActions.changePageTitle,
      changePageTitleButton: globalActions.changePageTitleButton,
      showListTitle: globalActions.toggleDetailTitleOff,
      changePageTarget: globalActions.changePageTarget,
      searchVal: globalActions.getSearchVal
    },
    dispatch
  );
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LocationsList);
