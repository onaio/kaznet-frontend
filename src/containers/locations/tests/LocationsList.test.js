import React from "react";
import { shallow, mount } from "enzyme";
import toJson from "enzyme-to-json";
import { Router } from "react-router";
import createBrowserHistory from "history/createBrowserHistory";

import { LocationsList } from "../LocationsList";
import * as fixtures from "../../../store/locations/tests/fixtures";
const history = createBrowserHistory();

describe("containers/locations/LocationList", () => {
  it("renders without crashing", () => {
    shallow(
      <LocationsList
        fetchLocations={function() {}}
        changePageTitle={function() {}}
        changePageTitleButton={function() {}}
        showListTitle={function() {}}
      />
    );
  });

  it("renders location list correctly", () => {
    const wrapper = mount(
      <Router history={history}>
        <LocationsList
          fetchLocations={function() {}}
          changePageTitle={function() {}}
          changePageTitleButton={function() {}}
          showListTitle={function() {}}
          rowsById={fixtures.locationById}
          rowsIdArray={fixtures.locationIdArray}
          pageLinks={fixtures.locationData.links}
          currentPage={fixtures.locationData.meta.pagination.page}
        />
      </Router>
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.unmount();
  });
});
