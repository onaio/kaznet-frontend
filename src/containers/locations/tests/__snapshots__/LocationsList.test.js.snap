// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`containers/locations/LocationList renders location list correctly 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "createHref": [Function],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <LocationsList
    changePageTitle={[Function]}
    changePageTitleButton={[Function]}
    currentPage={1}
    endpoint="locations"
    fetchLocations={[Function]}
    pageLinks={
      Object {
        "first": "http://localhost:8000/api/v1/locations/?page=1",
        "last": "http://localhost:8000/api/v1/locations/?page=1",
        "next": null,
        "prev": null,
      }
    }
    rowsById={
      Object {
        "1": Object {
          "attributes": Object {
            "country": "KE",
            "created": "2018-06-20T12:09:02.916318+03:00",
            "description": "Such Big City , Much wow",
            "geopoint": null,
            "location_type_name": "None",
            "modified": "2018-06-20T12:09:02.916345+03:00",
            "name": "Nairobi",
            "parent_name": "None",
            "radius": null,
            "shapefile": Object {
              "coordinates": Array [
                Array [
                  Array [
                    Array [
                      -0.038795471191406,
                      0.030555723649271,
                    ],
                    Array [
                      0.049095153808594,
                      0.013389587280456,
                    ],
                    Array [
                      0.015106201171875,
                      -0.063171373920095,
                    ],
                    Array [
                      -0.044288635253906,
                      -0.032615659859611,
                    ],
                    Array [
                      -0.038795471191406,
                      0.030555723649271,
                    ],
                  ],
                ],
              ],
              "type": "MultiPolygon",
            },
          },
          "id": "1",
          "relationships": Object {
            "location_type": Object {
              "data": null,
            },
            "parent": Object {
              "data": null,
            },
          },
          "type": "Location",
        },
      }
    }
    rowsIdArray={
      Array [
        "1",
      ]
    }
    showListTitle={[Function]}
  >
    <div
      className="LocationList"
    >
      <ListView
        currentPage={1}
        endpoint="locations"
        pageLinks={
          Object {
            "first": "http://localhost:8000/api/v1/locations/?page=1",
            "last": "http://localhost:8000/api/v1/locations/?page=1",
            "next": null,
            "prev": null,
          }
        }
        renderHeaders={[Function]}
        renderRow={[Function]}
        rowsById={
          Object {
            "1": Object {
              "attributes": Object {
                "country": "KE",
                "created": "2018-06-20T12:09:02.916318+03:00",
                "description": "Such Big City , Much wow",
                "geopoint": null,
                "location_type_name": "None",
                "modified": "2018-06-20T12:09:02.916345+03:00",
                "name": "Nairobi",
                "parent_name": "None",
                "radius": null,
                "shapefile": Object {
                  "coordinates": Array [
                    Array [
                      Array [
                        Array [
                          -0.038795471191406,
                          0.030555723649271,
                        ],
                        Array [
                          0.049095153808594,
                          0.013389587280456,
                        ],
                        Array [
                          0.015106201171875,
                          -0.063171373920095,
                        ],
                        Array [
                          -0.044288635253906,
                          -0.032615659859611,
                        ],
                        Array [
                          -0.038795471191406,
                          0.030555723649271,
                        ],
                      ],
                    ],
                  ],
                  "type": "MultiPolygon",
                },
              },
              "id": "1",
              "relationships": Object {
                "location_type": Object {
                  "data": null,
                },
                "parent": Object {
                  "data": null,
                },
              },
              "type": "Location",
            },
          }
        }
        rowsIdArray={
          Array [
            "1",
          ]
        }
      >
        <div>
          <Table
            bordered={true}
            className="kaznet-table"
            responsiveTag="div"
            tag="table"
          >
            <table
              className="kaznet-table table table-bordered"
            >
              <thead>
                <tr>
                  <ElementMap
                    HTMLTag="th"
                    items={
                      Array [
                        "Name",
                        "Parent",
                        "Type",
                        "Description",
                      ]
                    }
                  >
                    <th
                      key="0"
                    >
                      Name
                    </th>
                    <th
                      key="1"
                    >
                      Parent
                    </th>
                    <th
                      key="2"
                    >
                      Type
                    </th>
                    <th
                      key="3"
                    >
                      Description
                    </th>
                  </ElementMap>
                </tr>
              </thead>
              <tbody>
                <tr
                  key="1"
                >
                  <ElementMap
                    HTMLTag="td"
                    items={
                      Array [
                        "Nairobi",
                        "None",
                        "None",
                        "Such Big City , Much wow",
                      ]
                    }
                  >
                    <td
                      key="0"
                    >
                      Nairobi
                    </td>
                    <td
                      key="1"
                    >
                      None
                    </td>
                    <td
                      key="2"
                    >
                      None
                    </td>
                    <td
                      key="3"
                    >
                      Such Big City , Much wow
                    </td>
                  </ElementMap>
                </tr>
              </tbody>
            </table>
          </Table>
          <Pagination
            aria-label="Page navigation example"
            listTag="ul"
            tag="nav"
          >
            <nav
              aria-label="Page navigation example"
              className=""
            >
              <ul
                className="pagination"
              >
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="First"
                      className="page-link"
                      replace={false}
                      to="/locations/?page=first"
                    >
                      <a
                        aria-label="First"
                        className="page-link"
                        href="/locations/?page=first"
                        onClick={[Function]}
                      >
                        First
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={true}
                  tag="li"
                >
                  <li
                    className="page-item disabled"
                  >
                    <Link
                      aria-label="Previous"
                      className="page-link"
                      replace={false}
                      to="#"
                    >
                      <a
                        aria-label="Previous"
                        className="page-link"
                        href="/"
                        onClick={[Function]}
                      >
                        «
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={true}
                  tag="li"
                >
                  <li
                    className="page-item disabled"
                  >
                    <Link
                      aria-label="Next"
                      className="page-link"
                      replace={false}
                      to="#"
                    >
                      <a
                        aria-label="Next"
                        className="page-link"
                        href="/"
                        onClick={[Function]}
                      >
                        »
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
                <PaginationItem
                  disabled={false}
                  tag="li"
                >
                  <li
                    className="page-item"
                  >
                    <Link
                      aria-label="Last"
                      className="page-link"
                      replace={false}
                      to="/locations/?page=last"
                    >
                      <a
                        aria-label="Last"
                        className="page-link"
                        href="/locations/?page=last"
                        onClick={[Function]}
                      >
                        Last
                      </a>
                    </Link>
                  </li>
                </PaginationItem>
              </ul>
            </nav>
          </Pagination>
        </div>
      </ListView>
    </div>
  </LocationsList>
</Router>
`;
